### PR TITLE
[Kernel] DP + EP : GPTOSS +  DeepEP-HighThroughput

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
## Purpose
Integrate GPTOSS with DeepEPHTPrepareFinalize

Commands: 
server command: `VLLM_ALL2ALL_BACKEND="deepep_high_throughput" vllm serve openai/gpt-oss-20b  --port 9010  --data-parallel-size 2 --enable-expert-parallel  --no-enable-prefix-caching`
lm_eval command: `lm_eval --model local-completions --tasks gsm8k --model_args model=openai/gpt-oss-20b,base_url=http://127.0.0.1:9010/v1/completions,num_concurrent=30,max_retries=3 --limit 100`

Issue: The server some times  hangs / reports IMA. When the server runs through the lm_eval outputs are good. They look like 
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.31|±  |0.0465|
|     |       |strict-match    |     5|exact_match|↑  | 0.24|±  |0.0429|
```
and match main TP.

Debugging:
This PR uses `trtllm_fp4_block_scale_routed_moe` from flashinfer. I narrowed the issue down to the flashinfer kernel. 
1. One issue is that [link](https://github.com/flashinfer-ai/flashinfer/blob/cba02c4e9858eba08d26f97436e8fb42870492ef/csrc/trtllm_fused_moe_routing_renormalize.cu#L23) needs to be 256, otherwise it'll blow up at [link](https://github.com/flashinfer-ai/flashinfer/blob/cba02c4e9858eba08d26f97436e8fb42870492ef/csrc/trtllm_fused_moe_routing_renormalize.cu#L275) as `mPtrExpertCounts` is not big enough. 
I am still debugging this.

## (Optional) Documentation Update

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by zyongye** [comment]
Also a side note, please use the evaluation strategy in the [recipe](https://docs.vllm.ai/projects/recipes/en/latest/OpenAI/GPT-OSS.html#accuracy-evaluation-panels) instead of `lm_eval` for this model.

File: vllm/model_executor/layers/quantization/mxfp4.py [hunk 1]
```diff
@@ -373,6 +375,188 @@ def _get_tile_tokens_dim(self, x: torch.Tensor, top_k: int):
...
+            "output2_scale_scalar": None,
+            "num_experts": self.num_experts,
+            "top_k": top_k,
+            "n_group": None,
+            "topk_group": None,
+            "intermediate_size": self.intermediate_size,
+            "local_expert_offset": 0,
+            "local_num_experts": self.num_experts,
+            "routed_scaling_factor": None,
+            "tile_tokens_dim": 8,
```
Comment (IwakuraRein) [inline vllm/model_executor/layers/quantization/mxfp4.py:487]:
```suggestion
            "tile_tokens_dim": _get_tile_tokens_dim(x_quant, top_k, self.num_experts),
```

File: vllm/model_executor/layers/quantization/mxfp4.py [hunk 2]
```diff
@@ -373,6 +375,188 @@ def _get_tile_tokens_dim(self, x: torch.Tensor, top_k: int):
...
+            e_score_correction_bias: Optional[torch.Tensor] = None,
+            apply_router_weight_on_input: bool = False,
+            activation: str = "silu",
+            enable_eplb: bool = False,
+            expert_load_view: Optional[torch.Tensor] = None,
+            logical_to_physical_map: Optional[torch.Tensor] = None,
+            logical_replica_count: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+
+        topk_weights, topk_ids = FusedMoE.select_experts(
```
Comment (IwakuraRein) [inline vllm/model_executor/layers/quantization/mxfp4.py:432]: `topk_ids` and `topk_weights` needs to be local and non local experts' id should be -1.

Or use global `topk_ids` and `topk_weights`, and provide `local_expert_offset` and `local_num_experts`.

File: vllm/model_executor/layers/fused_moe/trtllm_moe.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            intermediate_size,
+            "local_expert_offset":
+            local_expert_offset,
+            "local_num_experts":
+            local_num_experts,
+            "routed_scaling_factor":
+            None,
+            "tile_tokens_dim":
+            self._get_tile_tokens_dim(x_quant, topk, local_num_experts),
+            "routing_method_type":
```
Comment (IwakuraRein) [inline vllm/model_executor/layers/fused_moe/trtllm_moe.py:184]: `routing_method_type` is hardcoded to renormalize. Maybe add assertion above to make sure it's not using a different routing method.

File: vllm/model_executor/layers/quantization/mxfp4.py (hunk 2 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/quantization/mxfp4.py:432]: The topk_ids and topk_weights gets processed into, 
> use global topk_ids and topk_weights, and provide local_expert_offset and local_num_experts
in the all2alls. I verified that this is correct.

**Comment by varun-sundar-rabindranath** [comment]
Update on debugging: 
 - I got `VLLM_USE_FLASHINFER_MOE_MXFP4_BF16` working reliably by explicitly adding the `--enforce-eager` option.
 - `VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8` still locks up. I have narrowed it down to `mPermuteGemm1.run` but afaict all the setup is correct and the permute indices inside is also computed correctly. looking into this further. 
- hunch: I think the interaction with torch compile might have something to do with this.

File: vllm/model_executor/layers/quantization/utils/mxfp8_utils.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+def mxfp8_quantize(x: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
+
+    try:
+        from flashinfer import mxfp8_quantize
+    except ImportError as err:
+        raise ImportError("The package `flashinfer` is required to do "
+                          "MX-FP8 quantization. Please install it with" \
+                          "`pip install flashinfer`") from err
+
+    return mxfp8_quantize(x, is_sf_swizzled_layout=False)
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/quantization/utils/mxfp8_utils.py:18]: @IwakuraRein  is this the right way to quantize bf16 activations to fp8 ? Thanks.

Comment (IwakuraRein) [reply vllm/model_executor/layers/quantization/utils/mxfp8_utils.py:18]: Yes, when you are using mxfp4 x mxfp8 path.

**Comment by IwakuraRein** [comment]
> * hunch: I think the interaction with torch compile might have something to do with this.

If torch.compile is using garbage values to initialize `routing_logits` and/or `topk_ids` then there might be Illegal memory access.
